From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Matteo Collina <hello@matteocollina.com>
Date: Wed, 1 Jul 2026 15:59:54 +0200
Subject: [PATCH 26/74] http: reject requests exceeding max header count

(cherry picked from commit 0d3139ce8c35ac866ee77e116833e4a2b1eeea11)

PR-URL: https://github.com/nodejs-private/node-private/pull/932
CVE-ID: CVE-2026-58044
(cherry picked from commit c8525ac3a6974f1db795d7fb5ac63e9f97a6fba5)
(cherry picked from commit 1a7ff1302a09e25414931755fca4c3e941fed982)
---
 src/node_http_parser.cc                       | 38 ++++++++++++
 .../test-http-max-headers-count-overflow.js   | 34 +++++++++++
 test/parallel/test-http-max-headers-count.js  | 58 +++++++++++--------
 test/parallel/test-https-max-headers-count.js | 40 ++++++++-----
 4 files changed, 133 insertions(+), 37 deletions(-)
 create mode 100644 test/parallel/test-http-max-headers-count-overflow.js

diff --git a/src/node_http_parser.cc b/src/node_http_parser.cc
index 99712a87c0..57215bb87b 100644
--- a/src/node_http_parser.cc
+++ b/src/node_http_parser.cc
@@ -329,6 +329,11 @@ class Parser : public AsyncWrap, public StreamListener {
 
     if (num_fields_ == num_values_) {
       // start of new field name
+      rv = TrackHeaderPair();
+      if (rv != 0) {
+        return rv;
+      }
+
       num_fields_++;
       if (num_fields_ == kMaxHeaderFieldsCount) {
         // ran out of space - flush to javascript land
@@ -413,6 +418,7 @@ class Parser : public AsyncWrap, public StreamListener {
 
     num_fields_ = 0;
     num_values_ = 0;
+    header_pairs_ = 0;
 
     // METHOD
     if (parser_.type == HTTP_REQUEST) {
@@ -506,6 +512,8 @@ class Parser : public AsyncWrap, public StreamListener {
     if (num_fields_)
       Flush();  // Flush trailing HTTP headers.
 
+    header_pairs_ = 0;
+
     Local<Object> obj = object();
     Local<Value> cb = obj->Get(env()->context(),
                                kOnMessageComplete).ToLocalChecked();
@@ -987,6 +995,7 @@ class Parser : public AsyncWrap, public StreamListener {
     got_exception_ = false;
     headers_completed_ = false;
     max_http_header_size_ = max_http_header_size;
+    header_pairs_ = 0;
   }
 
 
@@ -999,6 +1008,34 @@ class Parser : public AsyncWrap, public StreamListener {
     return 0;
   }
 
+  int TrackHeaderPair() {
+    if (parser_.type != HTTP_REQUEST) {
+      return 0;
+    }
+
+    header_pairs_ += 2;
+
+    Local<Value> max_header_pairs_v;
+    if (!object()
+             ->Get(env()->context(),
+                   FIXED_ONE_BYTE_STRING(env()->isolate(), "maxHeaderPairs"))
+             .ToLocal(&max_header_pairs_v)) {
+      got_exception_ = true;
+      return -1;
+    }
+
+    if (!max_header_pairs_v->IsNumber()) {
+      return 0;
+    }
+
+    const double max_header_pairs = max_header_pairs_v.As<Number>()->Value();
+    if (max_header_pairs > 0 && header_pairs_ > max_header_pairs) {
+      llhttp_set_error_reason(&parser_, "HPE_HEADER_OVERFLOW:Header overflow");
+      return HPE_USER;
+    }
+
+    return 0;
+  }
 
   int MaybePause() {
     if (!pending_pause_) {
@@ -1031,6 +1068,7 @@ class Parser : public AsyncWrap, public StreamListener {
   size_t current_buffer_len_;
   const char* current_buffer_data_;
   bool headers_completed_ = false;
+  size_t header_pairs_ = 0;
   bool pending_pause_ = false;
   uint64_t header_nread_ = 0;
   uint64_t chunk_extensions_nread_ = 0;
diff --git a/test/parallel/test-http-max-headers-count-overflow.js b/test/parallel/test-http-max-headers-count-overflow.js
new file mode 100644
index 0000000000..695195bba3
--- /dev/null
+++ b/test/parallel/test-http-max-headers-count-overflow.js
@@ -0,0 +1,34 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const http = require('http');
+const net = require('net');
+
+const server = http.createServer(common.mustNotCall());
+
+server.maxHeadersCount = 2;
+
+server.on('clientError', common.mustCall((err, socket) => {
+  assert.strictEqual(err.code, 'HPE_HEADER_OVERFLOW');
+  socket.end('HTTP/1.1 431 Request Header Fields Too Large\r\n\r\n');
+}));
+
+server.listen(0, common.mustCall(() => {
+  const port = server.address().port;
+  const req = 'POST / HTTP/1.1\r\n' +
+              'Host: localhost\r\n' +
+              'X-A: b\r\n' +
+              'Content-Length: 3\r\n' +
+              '\r\nabc';
+
+  net.createConnection(port, 'localhost', common.mustCall(function() {
+    let response = '';
+    this.setEncoding('latin1');
+    this.end(req);
+    this.on('data', (chunk) => response += chunk);
+    this.on('end', common.mustCall(() => {
+      assert.match(response, /^HTTP\/1\.1 431 /);
+      server.close();
+    }));
+  }));
+}));
diff --git a/test/parallel/test-http-max-headers-count.js b/test/parallel/test-http-max-headers-count.js
index b707c62bfb..13cb487d48 100644
--- a/test/parallel/test-http-max-headers-count.js
+++ b/test/parallel/test-http-max-headers-count.js
@@ -20,7 +20,7 @@
 // USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 'use strict';
-require('../common');
+const common = require('../common');
 const assert = require('assert');
 const http = require('http');
 
@@ -33,28 +33,38 @@ for (let i = 0; i < N; ++i) {
   headers[`key${i}`] = i;
 }
 
-const maxAndExpected = [ // for server
-  [50, 50],
-  [1500, 102],
-  [0, N + 2], // Host and Connection
+function createRequestHeaders(count) {
+  const requestHeaders = {
+    host: 'example.com',
+  };
+  for (let i = 0; i < count; ++i) {
+    requestHeaders[`key${i}`] = i;
+  }
+  return requestHeaders;
+}
+
+const serverMaxAndExpected = [ // for server
+  [50, 50, 48],
+  [1500, 102, N],
+  [0, N + 2, N], // Host and Connection
 ];
-let max = maxAndExpected[requests][0];
-let expected = maxAndExpected[requests][1];
+let max = serverMaxAndExpected[requests][0];
+let expected = serverMaxAndExpected[requests][1];
 
-const server = http.createServer(function(req, res) {
+const server = http.createServer(common.mustCallAtLeast((req, res) => {
   assert.strictEqual(Object.keys(req.headers).length, expected);
-  if (++requests < maxAndExpected.length) {
-    max = maxAndExpected[requests][0];
-    expected = maxAndExpected[requests][1];
+  if (++requests < serverMaxAndExpected.length) {
+    max = serverMaxAndExpected[requests][0];
+    expected = serverMaxAndExpected[requests][1];
     server.maxHeadersCount = max;
   }
   res.writeHead(200, headers);
   res.end();
-});
+}));
 server.maxHeadersCount = max;
 
-server.listen(0, function() {
-  const maxAndExpected = [ // for client
+server.listen(0, common.mustCall(() => {
+  const clientMaxAndExpected = [ // for client
     [20, 20],
     [1200, 103],
     [0, N + 3], // Connection, Date and Transfer-Encoding
@@ -62,28 +72,30 @@ server.listen(0, function() {
   doRequest();
 
   function doRequest() {
-    const max = maxAndExpected[responses][0];
-    const expected = maxAndExpected[responses][1];
+    const max = clientMaxAndExpected[responses][0];
+    const expected = clientMaxAndExpected[responses][1];
+    const requestHeaders =
+      createRequestHeaders(serverMaxAndExpected[requests][2]);
     const req = http.request({
       port: server.address().port,
-      headers: headers
-    }, function(res) {
+      headers: requestHeaders
+    }, common.mustCall((res) => {
       assert.strictEqual(Object.keys(res.headers).length, expected);
       res.on('end', function() {
-        if (++responses < maxAndExpected.length) {
+        if (++responses < clientMaxAndExpected.length) {
           doRequest();
         } else {
           server.close();
         }
       });
       res.resume();
-    });
+    }));
     req.maxHeadersCount = max;
     req.end();
   }
-});
+}));
 
 process.on('exit', function() {
-  assert.strictEqual(requests, maxAndExpected.length);
-  assert.strictEqual(responses, maxAndExpected.length);
+  assert.strictEqual(requests, serverMaxAndExpected.length);
+  assert.strictEqual(responses, serverMaxAndExpected.length);
 });
diff --git a/test/parallel/test-https-max-headers-count.js b/test/parallel/test-https-max-headers-count.js
index f8e4e64bfb..9cfc50ac87 100644
--- a/test/parallel/test-https-max-headers-count.js
+++ b/test/parallel/test-https-max-headers-count.js
@@ -22,19 +22,29 @@ for (let i = 0; i < N; ++i) {
   headers[`key${i}`] = i;
 }
 
-const maxAndExpected = [ // for server
-  [50, 50],
-  [1500, 102],
-  [0, N + 2], // Host and Connection
+function createRequestHeaders(count) {
+  const requestHeaders = {
+    host: 'example.com',
+  };
+  for (let i = 0; i < count; ++i) {
+    requestHeaders[`key${i}`] = i;
+  }
+  return requestHeaders;
+}
+
+const serverMaxAndExpected = [ // for server
+  [50, 50, 48],
+  [1500, 102, N],
+  [0, N + 2, N], // Host and Connection
 ];
-let max = maxAndExpected[requests][0];
-let expected = maxAndExpected[requests][1];
+let max = serverMaxAndExpected[requests][0];
+let expected = serverMaxAndExpected[requests][1];
 
 const server = https.createServer(serverOptions, common.mustCall((req, res) => {
   assert.strictEqual(Object.keys(req.headers).length, expected);
-  if (++requests < maxAndExpected.length) {
-    max = maxAndExpected[requests][0];
-    expected = maxAndExpected[requests][1];
+  if (++requests < serverMaxAndExpected.length) {
+    max = serverMaxAndExpected[requests][0];
+    expected = serverMaxAndExpected[requests][1];
     server.maxHeadersCount = max;
   }
   res.writeHead(200, headers);
@@ -43,22 +53,24 @@ const server = https.createServer(serverOptions, common.mustCall((req, res) => {
 server.maxHeadersCount = max;
 
 server.listen(0, common.mustCall(() => {
-  const maxAndExpected = [ // for client
+  const clientMaxAndExpected = [ // for client
     [20, 20],
     [1200, 103],
     [0, N + 3], // Connection, Date and Transfer-Encoding
   ];
   const doRequest = common.mustCall(() => {
-    const max = maxAndExpected[responses][0];
-    const expected = maxAndExpected[responses][1];
+    const max = clientMaxAndExpected[responses][0];
+    const expected = clientMaxAndExpected[responses][1];
+    const requestHeaders =
+      createRequestHeaders(serverMaxAndExpected[requests][2]);
     const req = https.request({
       port: server.address().port,
-      headers: headers,
+      headers: requestHeaders,
       rejectUnauthorized: false
     }, (res) => {
       assert.strictEqual(Object.keys(res.headers).length, expected);
       res.on('end', () => {
-        if (++responses < maxAndExpected.length) {
+        if (++responses < clientMaxAndExpected.length) {
           doRequest();
         } else {
           server.close();
